mirascope/core/bedrock/_utils/_setup_call.py

```diff
@@ -0,0 +1,156 @@
+    BedrockCallKwargs,
+]: ...
+
+
+@overload
+def setup_call(
+    *,
+    model: str,
+    client: BedrockRuntimeClient | AsyncBedrockRuntimeClient | None,
+    fn: Callable[..., BedrockDynamicConfig],
+    fn_args: dict[str, Any],
+    dynamic_config: BedrockDynamicConfig,
+    tools: list[type[BaseTool] | Callable] | None,
+    json_mode: bool,
+    call_params: BedrockCallParams,
+    extract: bool,
+) -> tuple[
+    CreateFn[ConverseResponseTypeDef, StreamOutputChunk],
+    str | None,
+    list[BedrockMessageParam],
+    list[type[BedrockTool]] | None,
+    BedrockCallKwargs,
+]: ...
+
+
+def setup_call(
+    *,
+    model: str,
+    client: BedrockRuntimeClient | AsyncBedrockRuntimeClient | None,
+    fn: Callable[..., BedrockDynamicConfig]
+    | Callable[..., Awaitable[BedrockDynamicConfig]],
+    fn_args: dict[str, Any],
+    dynamic_config: BedrockDynamicConfig,
+    tools: list[type[BaseTool] | Callable] | None,
+    json_mode: bool,
+    call_params: BedrockCallParams,
+    extract: bool,
+) -> tuple[
+    AsyncCreateFn[AsyncConverseResponseTypeDef, AsyncStreamOutputChunk]
+    | CreateFn[ConverseResponseTypeDef, StreamOutputChunk],
+    str | None,
+    list[BedrockMessageParam],
+    list[type[BedrockTool]] | None,
+    BedrockCallKwargs,
+]:
+    prompt_template, messages, tool_types, base_call_kwargs = _utils.setup_call(
+        fn, fn_args, dynamic_config, tools, BedrockTool, call_params
+    )
+    call_kwargs = cast(BedrockCallKwargs, base_call_kwargs)
+    messages = cast(list[BaseMessageParam | BedrockMessageParam], messages)
+    messages = convert_message_params(messages)
+    if messages[0]["role"] == "system":
+        call_kwargs["system"] = [
+            {"text": text}
+            for c in messages.pop(0)["content"]
+            if (text := c.get("text"))
+        ]
+
+    call_kwargs_tools = call_kwargs.pop("tools", None)
+    if json_mode:
+        json_mode_content = _utils.json_mode_content(
+            tool_types[0] if tool_types else None
+        )
+        if "text" in messages[-1]["content"][-1]:
+            messages[-1]["content"][-1]["text"] += json_mode_content
+        else:
+            messages[-1]["content"] = [
+                *messages[-1]["content"],
+                {"text": json_mode_content},
+            ]
+
+    else:
+        if call_kwargs_tools:
+            call_kwargs["toolConfig"] = cast(
+                ToolConfigurationTypeDef, {"tools": call_kwargs_tools}
+            )
+        if extract:
+            assert tool_types, "At least one tool must be provided for extraction."
+            if "toolConfig" in call_kwargs:
+                call_kwargs["toolConfig"]["toolChoice"] = cast(
+                    ToolChoiceTypeDef, {"type": "tool", "name": tool_types[0]._name()}
+                )
+
+    call_kwargs |= cast(BedrockCallKwargs, {"modelId": model, "messages": messages})
+
+    if client is None:
+        client = (
+            asyncio.run(_get_async_client())
```

Comment:
hmm, i think these refactored utilities are a little overkill. we only use them here, correct? we should just do this here in an if/else and call it a day, refactoring this without a need to reuse it is overkill

Reply:
something like this should work:

```python
# defined above so it's not defined on every call
def _get_async_client(session: ...) -> AsyncBedrockRuntimeClient:
    async with session.create_client("bedrock_runtime") as client:
        return client

# inside function
if client is None:
    if inspect.iscoroutinefunction(fn):
        session = get_session()
        # we need to get the async client through an async context manager
        # so we define a function to do this and run it here with asyncio
        client = asyncio.run(_get_async_client(session))
    else:
        session = Session()
        client = session.client("bedrock-runtime")
```